python/lsst/ci/build.py

```diff
@@ -345,6 +393,144 @@ def run(args):
+                )
+
+            # Attempt the build
+            retcode = b.build()
+
+        except UnknownAgentError as agent_err:
+            print(f"WARNING: {agent_err}")
+            retcode = b.build()
+
+        except Exception as other_ex:
+            print(f"Build failed on {agent}: {other_ex}")
+            traceback.print_exc()
+            retcode = False
+
+        finally:
+            # Ensure status is always written
+            b.write_status()
+
+            # Final status posting based on retcode
+            if retcode:
+                if pr_info.items:
+                    post_statuses.post_all_status(
+                        pr_info, agent=agent, state="success", description=f"Build succeeded on {agent}"
+                    )
+                else:
+                    print(f"Build succeeded on {agent}.")
+
+                sys.exit(0)
+            else:
+                if pr_info.items:
+                    post_statuses.post_all_status(
+                        pr_info, agent=agent, state="failure", description=f"Build failed on {agent}"
+                    )
+                else:
+                    print(f"Build failed on {agent}")
+                sys.exit(1)
+
+
+class UnknownAgentError(Exception):
+    """
+    Raised when the agent is unknown.
+
+    Parameters
+    ----------
+    agent : str
+        The agent that could not be determined.
+    """
+
+    def __init__(self, agent: str) -> None:
+        super().__init__(f"Unknown agent '{agent}', ensure sys.platform is available. Continuing build.")
+        self.agent = agent
+
+
+class PostStatuses:
+    """
+    Class that posts Github Statuses.
+
+    Parameters
+    ----------
+    pr_info
+        pydantic model including information on all open PRs
+    state
+        the state of the Github status
+    description
+        description of the current status
+    agent
+        current agent that is running
+    """
+
+    def post_github_status(
+        self, pr_info: AllProductPRCollections, state: str, description: str, agent: str
+    ) -> None:
+        """
+        Post a GitHub status to the commit in a specific PR.
+        """
+        token = os.getenv("GITHUB_TOKEN")
+        if not token:
+            print("GITHUB_TOKEN not found in environment variables.")
+            return
+
+        owner = pr_info.owner
+        repo = pr_info.repo
+        sha = pr_info.sha
+
+        url = f"https://api.github.com/repos/{owner}/{repo}/statuses/{sha}"
+        headers = {
+            "Authorization": f"token {token}",
+            "Accept": "application/vnd.github.v3+json",
+        }
+
+        print(f"Posting GitHub status to ({owner}/{repo}): {state} - {description}")
+
+        build_url = os.getenv(
+            "RUN_DISPLAY_URL",
+            "https://rubin-ci-dev.slac.stanford.edu/blue/organizations/jenkins/stack-os-matrix/activity",
```

Comment:
Not sure this is a good default value.  May be better to leave with no default, or at least one that's deployment-specific.